File: gaios/Models/CountlyWidget.swift
import Foundation

enum CountlyWidgetType: String {
    case nps
    case survey
    case undefined
}

enum SurveyFollowUpType: String {
    case score
    case one
    case none
}

enum WidgetQuestionType: String {
    case rating
    case text
    case undefined
}

struct CountlyWidget: Decodable {
    var _id: String?
    var app_id: String?
    var name: String?
    var msg: WidgetMessage?
    var appearance: WidgetAppearance?
    var type: String?
    var followUpType: String?
    var questions: [WidgetQuestion]?

    static func build(_ widget: [AnyHashable: Any]) -> CountlyWidget? {
        let json = try? JSONSerialization.data(withJSONObject: widget, options: [])
        let w = try? JSONDecoder().decode(CountlyWidget.self, from: json ?? Data())
        return w
    }

    var wType: CountlyWidgetType {
        guard let value = CountlyWidgetType(rawValue: self.type ?? "") else {
            return .undefined
        }
        return value
    }

    var wFollowUpType: SurveyFollowUpType {
        guard let value = SurveyFollowUpType(rawValue: self.followUpType ?? "") else {
            return .none
        }
        return value
    }
}

struct WidgetMessage: Decodable {
    var mainQuestion: String?
    var followUpAll: String?
    var followUpPromoter: String?
    var followUpPassive: String?
    var followUpDetractor: String?
    var thanks: String?
}

struct WidgetAppearance: Decodable {
    var show: String?
    var color: String?
    var style: String?
    var submit: String?
    var followUpInput: String?
    var notLikely: String?
    var likely: String?
}

struct WidgetQuestion: Decodable {
    var type: String?
    var question: String?
    var required: Bool?
    var other: Bool?
    var allOfTheAbove: Bool?
    var noneOfTheAbove: Bool?
    var otherText: String?
    var allOfTheAboveText: String?
    var noneOfTheAboveText: String?
    var notLikely: String?
    var likely: String?
    var id: String?

    var qType: WidgetQuestionType {
        guard let value = WidgetQuestionType(rawValue: self.type ?? "") else {
            return .undefined
        }
        return value
    }
}

File: common/src/commonMain/kotlin/com/blockstream/common/data/CountlyWidget.kt
package com.blockstream.common.data

import com.blockstream.common.gdk.GreenJson
import com.blockstream.common.serializers.HtmlEntitiesSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

@Serializable
data class CountlyWidget(
    @SerialName("_id") val id: String,
    @SerialName("app_id") val appId: String,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("name") val name: String,
    @SerialName("questions") val questions: List<Question>? = null,
    @SerialName("type") val type: SurveyType,
    @SerialName("msg") val msg: Messages,
    @SerialName("appearance") val appearance: Appearance,
    @SerialName("followUpType") val followUpType: FollowUpType? = null,
) : GreenJson<CountlyWidget>() {
    override fun keepJsonElement() = true

    override fun kSerializer() = serializer()

    @Transient
    lateinit var widget : Any

    val text: Question? get() = questions?.find { it.type == "text" }
    val rating: Question? get() = questions?.find { it.type == "rating" }
}


@Serializable
enum class FollowUpType{
    @SerialName("score") Score, @SerialName("one") One, @SerialName("none") None
}

@Serializable
enum class SurveyType{
    @SerialName("nps") NPS, @SerialName("survey") Survey
}

@Serializable
data class Messages(
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("thanks") val thanks: String,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("mainQuestion") val mainQuestion: String = "",
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("followUpAll") val followUpAll: String = "",
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("followUpPromoter") val followUpPromoter: String = "",
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("followUpPassive") val followUpPassive: String = "",
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("followUpDetractor") val followUpDetractor: String = "",
) : GreenJson<Messages>() {
    override fun kSerializer() = serializer()
}

@Serializable
data class Appearance(
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("followUpInput") val followUpInput: String? = null,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("notLikely") val notLikely: String? = null,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("likely") val likely: String? = null,
) : GreenJson<Appearance>() {
    override fun kSerializer() = serializer()
}

@Serializable
data class Question(
    @SerialName("id") val id: String,
    @SerialName("type") val type: String,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("question") val question: String,
    @SerialName("required") val required: Boolean = false,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("followUpInput") val followUpInput: String? = null,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("notLikely") val notLikely: String? = null,
    @Serializable(with = HtmlEntitiesSerializer::class)
    @SerialName("likely") val likely: String? = null,
) : GreenJson<Question>() {
    override fun kSerializer() = serializer()
}

